elastic-agent-client.proto

```diff
@@ -112,15 +115,361 @@ message ActionResponse {
+    // of two responses will always be returned. The last response has eof set.
+    rpc Fetch(ArtifactFetchRequest) returns (stream ArtifactFetchResponse);
+}
+
+// A single metric that is reported by a unit. A unit can report multiple metrics in the same period.
+message UnitMetric {
+    enum MetricType {
+        KEYWORD = 0;
+        LONG = 1;
+    }
+
+    // Label for the metric.
+    string label = 1;
+
+    // Type for the metric.
+    MetricType type = 2;
+
+    // Value for the metric.
+    bytes value = 3;
+}
+
+message UnitMetrics {
+    // ID of the unit.
+    string id = 1;
+
+    // Metrics for this unit in this period.
+    repeated UnitMetric metrics = 2;
+}
+
+message MetricsReportRequest {
+    // Token that is used to uniquely identify the collection of inputs to the agent. When started this is provided
+    // in the `ConnInfo`.
+    string token = 1;
+
+    // Period of time for the metrics that are collected.
+    uint64 period = 2;
+
+    // Reported units for this period.
+    repeated UnitMetrics units = 3;
+}
+
+// Response from `MetricsReport`.
+message MetricsReportResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+service ElasticAgentMetrics {
+    // Report metrics to the Elastic Agent.
+    rpc Report(MetricsReportRequest) returns (MetricsReportResponse);
+}
+
+message LogMessageRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // ID of the unit.
+    string unit_id = 2;
+    // ECS log message body JSON encoded.
+    bytes message = 3;
+}
+
+message LogMessageResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+message LogBulkMessage {
+    // ID of the unit.
+    string unit_id = 1;
+    // ECS log message body JSON encoded.
+    bytes message = 2;
+}
+
+message LogBulkMessageRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Multiple message to report at the same time.
+    repeated LogBulkMessage messages = 2;
+}
+
+message LogBulkMessageResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Log service is only exposed to programs that are not started as sub-processes by Elastic Agent.
+//
+// This allows services that are not started as sub-processes to write to the same stdout that programs that do get
+// started as subprocess to write to. A program that is as a sub-process with stdout connected does not have the
+// ability to use this service.
+service ElasticAgentLog {
+    // Log a message to the Elastic Agent.
+    rpc Log(LogMessageRequest) returns (LogMessageResponse);
+
+    // Bulk log messages to the Elastic Agent.
+    rpc LogBulk(LogBulkMessageRequest) returns (LogBulkMessageResponse);
+}
```

Comment:
Do we really need 2 APIs for that? One bulk API where it might have one or more messages isn't enough?

Reply:
Agree on this one, let's just use a `bulk`

Reply:
Updated.